### PR TITLE
Keep Core memory inside cygrpc.Channel objects

### PR DESCRIPTION
**Pull request by nathanielmanistaatgoogle**
At long last...

### CODE REVIEW

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi [hunk 1]
```diff
@@ -20,6 +20,59 @@ import time
...
+    return None, ConnectivityEvent(GRPC_QUEUE_SHUTDOWN, False, None)
+  else:
+    tag = <_Tag>c_event.tag
+    # We receive event tags only after they've been inc-ref'd elsewhere in
+    # the code.
+    cpython.Py_DECREF(tag)
+    return tag, tag.event(c_event)
+
+
+cdef _immediate_event(grpc_completion_queue *c_completion_queue):
```
Comment (kpayson64) [inline src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi]: I'm assuming this function is used to extract a completion we know will be on cq, available immediately.

Can we not use an infinite timeout?  Also should we release the GIL?

File: src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi [hunk 2]
```diff
@@ -14,82 +14,445 @@
 
 cimport cpython
 
+import collections
```
Comment (kpayson64) [inline src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi]: Seems to be unused.

File: src/python/grpcio/grpc/_cython/_cygrpc/channel.pxd.pxi [hunk 3]
```diff
@@ -13,9 +13,61 @@
...
+
+cdef _check_call_error(c_call_error, metadata)
+
+
+cdef class _CallState:
+
+  cdef grpc_call *c_call
+  # A collections.MutableSet of the _Tags associated with pending batch
+  # operations for the call.
+  cdef set due
```
[the remark this answers is not in the document]
Comment (nathanielmanistaatgoogle) [reply src/python/grpcio/grpc/_cython/_cygrpc/channel.pxd.pxi]: "Effective Java", Second Edition, Item 52: "Refer to objects by their interfaces". I've been spending a lot of quality time with pytype lately.

... but if Cython won't let me call it a `collections.MutableSet`, then I should just call it a `set` and file a Cython bug.

File: src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi (hunk 2 above)
Comment (nathanielmanistaatgoogle) [reply src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi]: Will remove.

File: src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi [hunk 4]
```diff
@@ -14,82 +14,445 @@
...
+
+
+cdef str _call_error(c_call_error, metadata):
+  if c_call_error == GRPC_CALL_ERROR_INVALID_METADATA:
+    return _call_error_metadata(metadata)
+  else:
+    return _call_error_no_metadata(c_call_error)
+
+
+cdef _check_call_error_no_metadata(c_call_error):
```
[the remark this answers is not in the document]
Comment (nathanielmanistaatgoogle) [reply src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi]: The "no" here means "metadata isn't involved, so don't bother exploring whether or not the error is an invalid metadata error". Would you like a comment here?

File: src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi [hunk 5]
```diff
@@ -14,82 +14,445 @@
...
+        grpc_call_cancel_with_status(
+            call_state.c_call, code, encoded_details, NULL)
+      # TODO(https://github.com/grpc/grpc/issues/3064): Cancel connectivity
+      # watching.
+
+      while state.integrated_call_states:
+        state.condition.wait()
+      while state.segregated_call_states:
+        state.condition.wait()
+      while state.connectivity_due:
```
[the remark this answers is not in the document]
Comment (nathanielmanistaatgoogle) [reply src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi]: [Yes](https://github.com/grpc/grpc/issues/3064); see the TODO a few lines earlier.

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi (hunk 1 above)
[the remark this answers is not in the document]
Comment (nathanielmanistaatgoogle) [reply src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi]: Probably!

**Comment by nathanielmanistaatgoogle** [comment]
Merging as a peek in the log of the still-in-progress Basic Tests MacOS [opt] (internal CI) shows Python having already run and passed.

**Pull request merged**